lib/Lower/Bridge.cpp

```diff
@@ -147,58 +125,56 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+              return c->getMpSubprogram();
+            },
+            [&](Fortran::lower::pft::Evaluation *c) {
+              return inMpSubprogram(c);
             },
-            [&](Fortran::lower::pft::Evaluation *c) { return inMPSubp(c); },
             [](auto *) -> const Fortran::parser::MpSubprogramStmt * {
               return nullptr;
             },
         },
-        cstr->parent.p);
+        cstr->parentType.p);
   }
 
-  template <typename A>
-  static const Fortran::semantics::SomeExpr *
-  getScalarExprOfTuple(const A &tuple) {
-    return Fortran::semantics::GetExpr(
-        std::get<Fortran::parser::ScalarLogicalExpr>(tuple));
+  void genFIRUnconditionalBranch(mlir::Block *targetBlock) {
+    assert(targetBlock && "missing unconditional target block");
+    builder->create<mlir::BranchOp>(toLocation(), targetBlock);
   }
-  template <typename A>
-  static const Fortran::semantics::SomeExpr *getExprOfTuple(const A &tuple) {
-    return Fortran::semantics::GetExpr(
-        std::get<Fortran::parser::LogicalExpr>(tuple));
-  }
-  /// Get the condition expression for a CondGoto evaluation
-  const Fortran::semantics::SomeExpr *
-  getEvaluationCondition(Fortran::lower::pft::Evaluation &eval) {
-    return eval.visit(Fortran::common::visitors{
-        [&](const Fortran::parser::IfStmt &stmt) {
-          return getScalarExprOfTuple(stmt.t);
-        },
-        [&](const Fortran::parser::IfThenStmt &stmt) {
-          return getScalarExprOfTuple(stmt.t);
-        },
-        [&](const Fortran::parser::ElseIfStmt &stmt) {
-          return getScalarExprOfTuple(stmt.t);
-        },
-        [&](const Fortran::parser::WhereConstructStmt &stmt) {
-          return getExprOfTuple(stmt.t);
-        },
-        [&](const Fortran::parser::MaskedElsewhereStmt &stmt) {
-          return getExprOfTuple(stmt.t);
-        },
-        [&](auto &) -> const Fortran::semantics::SomeExpr * {
-          mlir::emitError(toLocation(), "unexpected conditional branch case");
-          return nullptr;
-        },
-    });
+
+  void
+  genFIRUnconditionalBranch(Fortran::lower::pft::Evaluation *targetEvaluation) {
+    builder->create<mlir::BranchOp>(toLocation(), targetEvaluation->block);
+  }
+
+  void genFIRConditionalBranch(mlir::Value &cond, mlir::Block *trueTarget,
+                               mlir::Block *falseTarget) {
+    assert(trueTarget && "missing conditional branch true block");
+    assert(falseTarget && "missing conditional branch false block");
+    llvm::SmallVector<mlir::Value, 2> blk;
+    builder->create<mlir::CondBranchOp>(toLocation(), cond, trueTarget, blk,
+                                        falseTarget, blk);
+  }
+
+  void genFIRConditionalBranch(mlir::Value &cond,
+                               Fortran::lower::pft::Evaluation *trueTarget,
+                               Fortran::lower::pft::Evaluation *falseTarget) {
+    genFIRConditionalBranch(cond, trueTarget->block, falseTarget->block);
+  }
```

Comment:
It looks like you might be able to get rid of this helper function.

Reply:
Good catch.  I'm no longer invoking the middle of these three functions.